articles/130_ros_time.md

```diff
@@ -39,85 +45,110 @@ Slower than real time simulation is necessary for complicated systems where accu
 
 #### System Time
 
-For convenience in these cases we will also provide the same API as above, but use the name `SystemTime`.
+System Time pulls the current time from the system clock chip. 
+It provides benefit over Steady Time when all nodes in the system have synchronized system time available or when all nodes run on the same host that contains a system clock chip.
 
-`SystemTime` will be directly tied to the system clock.
+We expect a System Time publisher to throw an error if the system clock resource is not available.
+We expect methods returning the current time to fall back to Steady Time if System Time is not available.
 
 #### Steady Time
 
-Example use cases for this include hardware drivers which are interacting with peripherals with hardware timeouts.
+Steady Time implies a real-time incrementing source.
+There is no modern hardware or operating system that does not support this option.
+Steady Time typically begins at system bootup and measures increments using hardware frequency counters.
 
-In nodes which require the use of `SteadyTime` or `SystemTime` for interacting with hardware or other peripherals it is expected that they do a best effort to isolate any `SystemTime` or `SteadyTime` information inside their implementation and translate external interfaces to use the ROS time abstraction when communicating over the ROS network.
+Example use cases for this include hardware drivers which are interacting with peripherals with hardware timeouts and systems lacking a realtime clock chip.
+
+In nodes which require the use of `SteadyTime` or `SystemTime` for interacting with hardware or other peripherals, it is expected that they do a best effort to isolate any `SystemTime` or `SteadyTime` information inside their implementation and translate external interfaces to use the ROS time abstraction when communicating over the ROS network.
 
 #### ROS Time
 
-The `ROSTime` will report the same as `SystemTime` when an ROS Time Source is not active.
-When the ROS time source is active `ROSTime` will return the latest value reported by the Time Source.
+ROS Time derives the current time from the most-recently received clock broadcast message published on the `/clock` topic.
 
-### ROS Time Sources
+When the ROS Time source is active `ROSTime` will return the latest value reported by the Time Source plus the elapsed time since that report multiplied by the real time factor.
 
-#### Default Time Source
+The ROS Time broadcast (clock message) needs to include this information:
+- The current time (in nanoseconds, epoch depends upon source and should not be needed)
+- The current real time factor (also called the "multiplier", typically a value of one, but could be zero for paused or negative when going backwards).
+- The system clock time (in nanoseconds since the Unix epoch) when the message was created (or -1 if not available)
```

Comment:
I do not think we can assume the system clock is not available. Also, `-1` is not a special value in system time. It would just be 1 second before 1970 if the seconds field is signed. I'd recommend `0` be the go to value if system time is not yet available (still not sure that make sense).

This is the behavior in ROS 1 (btw):

```
% python
Python 2.7.11 (default, Jan 18 2016, 12:41:04)
[GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import rospy
>>> rospy.Time.now()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/william/indigo/install/lib/python2.7/site-packages/rospy/rostime.py", line 155, in now
    return get_rostime()
  File "/Users/william/indigo/install/lib/python2.7/site-packages/rospy/rostime.py", line 209, in get_rostime
    raise rospy.exceptions.ROSInitException("time is not initialized. Have you called init_node()?")
rospy.exceptions.ROSInitException: time is not initialized. Have you called init_node()?
>>> rospy.init_node('foo')
>>> rospy.Time.now()
rospy.Time[1493933142113349914]
>>> ^D

% rosparam set use_sim_time true

% python
Python 2.7.11 (default, Jan 18 2016, 12:41:04)
[GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import rospy
>>> rospy.Time.now()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/william/indigo/install/lib/python2.7/site-packages/rospy/rostime.py", line 155, in now
    return get_rostime()
  File "/Users/william/indigo/install/lib/python2.7/site-packages/rospy/rostime.py", line 209, in get_rostime
    raise rospy.exceptions.ROSInitException("time is not initialized. Have you called init_node()?")
rospy.exceptions.ROSInitException: time is not initialized. Have you called init_node()?
>>> rospy.init_node('foo')
>>> rospy.Time.now()
rospy.Time[0]
>>>
```

As you can see it also raises an exception if a node has not been created yet. If I then start a rosbag I will start getting some sim time (with a big initial jump from 0 to the first time in the bag):

```
>>> # started rosbag play with --clock
...
>>> rospy.Time.now()
rospy.Time[1481249541938720020]
```

Not to say that this is the only way to do it, but I think it works ok in ROS 1 and I don't see any alternatives with advantages atm.